### ISSUE TITLE
ResponderGuard defense and ideas

### ISSUE THREAD

**Issue by Kevin-Robertson** [CredDefense/CredDefense#4]
Great work on this. I added some defenses to the Inveigh dev repo. Basically, since the NBNS request destination IP is often not the broadcast IP, I can hook on that to detect ResponderGuard. Right now, I just have an ignore in for each request. The next step would be to blacklist the ResponderGuard IP upon detection.

The broadcast address will be hit with ResponderGuard -CidrRange. It gets hit last though giving me a chance to blacklist. Maybe start with the broadcast address so that ResponderGuard will detect Inveigh when they are on the same subnet? The Inveigh start time vs ResponderGuard start time would be a factor but it probably increases the odds a little on the ResponderGuard side. Note that it seems like Inveigh running in unpriv mode doesn't even see the ResponderGuard non-broadcast requests. It's listening directly on the broadcast address rather than sniffing UDP 137 traffic.

Two other current ResponderGuard hooks I spotted are the random non-137 source port and the fixed NBNS transaction ID.

Another feature that might be helpful would be the ability to set the requested hostname. Maybe even add the ability to loop though an array of specified hostnames. I think my sleep deprived DerbyCon brain did a bad job of trying to explain it when we spoke. If an attacker watches the traffic (advisable) before spoofing, they can possibly spot a detection tool or even just zero in on something specific to spoof. Often, LLMNR and WPAD are all that's needed when spoofing. Being able to throw something like WPAD into the mix will help on the ResponderGuard side. When spoofing I'm also instantly suspicious of a request for a single hostname that looks random.

I'd love to see LLMNR/NBNS lose their default enabled status on domain attached systems so I'm all for beating up on them. Any opinion though on if I should enable ResponderGuard defense by default or not with Inveigh? Ideally, spoofing detection tools would be employed by blue teams that are hunting or organizations that have an unavoidable (hrm) need for LLMNR/NBNS. Running detection tools instead of hardening can be asking for trouble but so can just running spoofing tools without planning.

**Comment by dafthack** [comment]
Hey man! Great seeing you at Derbycon. Awesome points made. I totally agree with not running detection tools in place of hardening. In a perfect world you would definitely want hardening first, but it still would be nice to go hunting. :-) I can definitely work on fixing most of the issues you called out. 

-The non-137 source port and the fixed NBNS transaction ID shouldn't be too hard to fix. I'll work on those first.
-Setting the specified hostname is something I want but with the way I'm generating a random hostname it needed to be the same size (me rage coding it to get it in PoC form). But again, that is something that really shouldn't take terribly long to fix. What do you think about by default setting the hostname to something like "SERVERXXXXX" where the X's are randomized? That might look a little less obvious. What do you think about an option that enumerates all AD computer objects, figures out which ones are no longer reachable, and uses those hostnames as a list?
-Both WPAD and LLMNR are things I want added so I totally agree with you on those. Adding to the list!
-The broadcast IP problem is hard for me to think how I would get around that... I will have to think about that some more... but I like what you said about starting by scanning broadcast.

Your question about whether enabling defense in Inveigh against ResponderGuard by default is an interesting one. I would say that would be similar to someone writing malware obfuscation asking an AV company if they should continue to keep bypassing them or suffer detection. I think the answer is yes, keep pushing on developing bypasses because real attackers are sure as hell going to. But... I also would say if you have time to make the detection side better by helping me code in better detection in ResponderGuard then that could be time very well spent as well. :-) Thanks for taking a look at the tool and providing your feedback so quickly! I really appreciate it.

**Comment by Kevin-Robertson** [comment]
Yep, it would be great to explore detection more. If you want we can leave this open and I'll add in anything I see.

I was thinking about the hostname prefix also. The big thing would be to make it settable (this is how I end up with all those parameters in Inveigh!) so that there isn't a default I can use to detect ResponderGuard. The AD lookup thing would be cool. You could also maybe mirror some of the real requests pulled from the ResponderGuard subnet.

There is probably some stuff in Inveigh/Conveigh that can be looked at. I wasn't really happy with Conveigh using an elevated shell due to the packet sniffing code. I ran into some timing issues with just using broadcast/multicast listeners. It's the same reason that I don't currently have the learning mode working on the unpriv side of Inveigh. So maybe you'll solve that on the way:) My dream with Conveigh was that an admin would take one look at all the LLMNR/NBNS traffic bouncing around and immediately disable them both. It would be great though to have a good detection tool that can be used across an entire enterprise without dealing with routers.

**Comment by green69** [comment]
Amazing job with ResponderGuard.  Its going to help me in the next Pentest.  
Here is a little line of code since you helped me out that you can add to send email alerts in your script on Line 449.
Send-MailMessage -From "ResponderGuard <ResponderGuard@server.com>" -To "Christopher Mulligan <christopher.mulligan.1@ulaval.ca>", 
-Subject "Responder Guard as detected an attacker" -Body "An NBNS spoofer was discovered at $ip." -Priority High -SmtpServer "smtp.hotmail.com"

Thanks,
Chris

**Comment by Kevin-Robertson** [comment]
Just an update, Inveigh 1.3.2 with the ResponderGuard detection stuff never made it out of the dev branch. It will be in the upcoming 1.4 release though. The plan is to just have a parameter that enables/disables (enabled by default) defenses against detectors.
